File: server/history.js
/**
 * 历史会话存储（JSON 文件持久化，无需数据库）
 *
 * 结构：data/history.json
 * {
 *   "<sessionId>": {
 *     "id": "...", "title": "...", "createdAt": 1234567890000,
 *     "updatedAt": 1234567890000, "messages": [{ "role": "user"|"assistant", "content": "..." }]
 *   }
 * }
 *
 * 提供：列表 / 读取 / upsert / 删除，并自动按 updatedAt 排序、裁剪超量会话。
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');
const DATA_DIR = process.env.DATA_DIR
  ? path.resolve(process.env.DATA_DIR)
  : path.join(ROOT, 'data');
const FILE = path.join(DATA_DIR, 'history.json');

const MAX_SESSIONS = Number(process.env.HISTORY_MAX_SESSIONS) > 0 ? Number(process.env.HISTORY_MAX_SESSIONS) : 200;
const MAX_MESSAGES = 200; // 每个会话最多保留的消息数
const MAX_TEXT = 8000; // 单条消息最长字符数
const MAX_TITLE = 24;

function load() {
  try {
    return JSON.parse(fs.readFileSync(FILE, 'utf8'));
  } catch {
    return {};
  }
}

function save(store) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
  const tmp = `${FILE}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(store, null, 2), 'utf8');
  fs.renameSync(tmp, FILE); // 原子替换，避免半截文件
}

function sanitizeMessages(messages) {
  if (!Array.isArray(messages)) return [];
  return messages
    .filter((m) => m && (m.role === 'user' || m.role === 'assistant') && typeof m.content === 'string')
    .map((m) => ({ role: m.role, content: m.content.slice(0, MAX_TEXT) }))
    .slice(-MAX_MESSAGES);
}

function titleFrom(messages) {
  const first = messages.find((m) => m.role === 'user');
  const t = first ? first.content.trim() : '新对话';
  return (t.replace(/\s+/g, ' ').slice(0, MAX_TITLE) || '新对话').trim();
}

/** 会话列表（按更新时间倒序，含标题/条数/时间） */
export function listHistory() {
  const store = load();
  const list = Object.values(store).map((s) => ({
    id: s.id,
    title: s.title,
    updatedAt: s.updatedAt,
    count: (s.messages || []).length,
  }));
  list.sort((a, b) => b.updatedAt - a.updatedAt);
  return list.slice(0, MAX_SESSIONS);
}

/** 读取单个会话（完整消息） */
export function getHistory(id) {
  const s = load()[id];
  if (!s) return null;
  return {
    id: s.id,
    title: s.title,
    createdAt: s.createdAt,
    updatedAt: s.updatedAt,
    messages: s.messages || [],
  };
}

/** 新建或更新会话，返回完整会话对象 */
export function saveHistory(id, messages, title) {
  if (!id || typeof id !== 'string') throw new Error('缺少会话 id');
  const store = load();
  const existing = store[id] || {};
  const now = Date.now();
  const msgs = sanitizeMessages(messages);
  const session = {
    id,
    title: (title && title.trim().slice(0, MAX_TITLE)) || existing.title || titleFrom(msgs) || '新对话',
    createdAt: existing.createdAt || now,
    updatedAt: now,
    messages: msgs,
  };
  store[id] = session;

  // 超出上限时按最久未更新裁剪
  if (Object.keys(store).length > MAX_SESSIONS) {
    const keys = Object.keys(store).sort((a, b) => store[a].updatedAt - store[b].updatedAt);
    keys.slice(0, Object.keys(store).length - MAX_SESSIONS).forEach((k) => delete store[k]);
  }

  save(store);
  return session;
}

/** 删除单个会话 */
export function deleteHistory(id) {
  const store = load();
  if (store[id]) {
    delete store[id];
    save(store);
    return true;
  }
  return false;
}

File: src/voice_chat/history.py
"""历史会话存储(JSON 文件持久化),与旧 Node 版 data/history.json 完全兼容。

文件结构:{"<sessionId>": {"id","title","createdAt","updatedAt","messages":[{role,content}]}}
列表按 updatedAt 倒序;超量按最久未更新裁剪;写入用临时文件原子替换。
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path

MAX_MESSAGES = 200
MAX_TEXT = 8000
MAX_TITLE = 24

_MESSAGE_ROLES = ("user", "assistant")


class HistoryStore:
    def __init__(self, *, data_dir: Path | None = None, max_sessions: int = 200) -> None:
        if data_dir is None:
            env_dir = os.environ.get("DATA_DIR")
            data_dir = Path(env_dir) if env_dir else Path(__file__).resolve().parents[2] / "data"
        self._data_dir = data_dir
        self._file = data_dir / "history.json"
        self._max_sessions = max_sessions if max_sessions > 0 else 200

    # ---------- 读取 ----------

    def _load(self) -> dict:
        try:
            raw = json.loads(self._file.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            return {}
        return raw if isinstance(raw, dict) else {}

    def _persist(self, store: dict) -> None:
        self._data_dir.mkdir(parents=True, exist_ok=True)
        tmp = self._file.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(store, ensure_ascii=False, indent=2), encoding="utf-8")
        os.replace(tmp, self._file)  # 原子替换,避免半截文件

    # ---------- 对外操作 ----------

    def list(self) -> list[dict]:
        items = [
            {
                "id": s.get("id"),
                "title": s.get("title"),
                "updatedAt": s.get("updatedAt"),
                "count": len(s.get("messages") or []),
            }
            for s in self._load().values()
            if isinstance(s, dict)
        ]
        items.sort(key=lambda x: x["updatedAt"] or 0, reverse=True)
        return items[: self._max_sessions]

    def get(self, session_id: str) -> dict | None:
        s = self._load().get(session_id)
        if not isinstance(s, dict):
            return None
        return {
            "id": s.get("id"),
            "title": s.get("title"),
            "createdAt": s.get("createdAt"),
            "updatedAt": s.get("updatedAt"),
            "messages": s.get("messages") or [],
        }

    def save(self, session_id: str, messages, title: str | None = None) -> dict:
        if not session_id or not isinstance(session_id, str):
            raise ValueError("缺少会话 id")
        store = self._load()
        existing = store.get(session_id) if isinstance(store.get(session_id), dict) else {}
        now = int(time.time() * 1000)
        msgs = _sanitize_messages(messages)
        session = {
            "id": session_id,
            "title": _pick_title(title, existing.get("title"), _title_from_messages(msgs)),
            "createdAt": existing.get("createdAt") or now,
            "updatedAt": now,
            "messages": msgs,
        }
        store[session_id] = session
        self._prune(store)
        self._persist(store)
        return session

    def delete(self, session_id: str) -> bool:
        store = self._load()
        if session_id in store:
            del store[session_id]
            self._persist(store)
            return True
        return False

    # ---------- 内部 ----------

    def _prune(self, store: dict) -> None:
        if len(store) <= self._max_sessions:
            return
        keys = sorted(store, key=lambda k: (store[k].get("updatedAt") or 0))
        for k in keys[: len(store) - self._max_sessions]:
            del store[k]


def _sanitize_messages(messages) -> list[dict]:
    if not isinstance(messages, list):
        return []
    out = []
    for m in messages:
        if not isinstance(m, dict):
            continue
        role, content = m.get("role"), m.get("content")
        if role not in _MESSAGE_ROLES or not isinstance(content, str):
            continue
        out.append({"role": role, "content": content[:MAX_TEXT]})
    return out[-MAX_MESSAGES:]


def _title_from_messages(msgs: list[dict]) -> str:
    first = next((m for m in msgs if m["role"] == "user"), None)
    text = " ".join(first["content"].split()) if first else ""
    return text


def _pick_title(*candidates) -> str:
    for c in candidates:
        if isinstance(c, str) and c.strip():
            return c.strip()[:MAX_TITLE]
    return "新对话"
